agent-protocol.md

```diff
@@ -0,0 +1,224 @@
+
+## SMP agent protocol components
+
+SMP agent protocol has 3 main parts:
+
+- the syntax and semantics of messages that SMP agents exchange between each other to negotiate establishing unidirectional (simplex) encrypted queues on SMP server(s) to provide their users a higher level interface than SMP protocol to establish and operate duplex (bi-directional) connections and notifications about any communication integrity violations (via hash validation, sequential message IDs and message acknowledgements send over SMP queues).
+- the syntax and semantics of the commands that should be sent over TCP or other sequential protocol by the agent user to the agent. This protocol allows to create and manage multiple connections, each consisting of two simplex SMP queues.
+- the syntax and semantics of the message that users of SMP agents should send out-of-band (as pre-shared key / "invitation") to ensure [E2E encryption][1] integrity for the first SMP queue and protection against active attacks ([MITM attacks][2]).
+
+## Duplex connection
+
+**Creating duplex connection between Alice and Bob:**
+
+![Creating duplex connection](/diagrams/duplex-messaging/duplex-creating.svg)
+
+## Communication between SMP agents
+
+SMP agents communicate via SMP servers managing creation, deletion and operations of SMP queues.
+
+Agents can use SMP message client body to transmit agent client messages and exchange messages between each other.
+
+Each SMP message client body, once decrypted, contains 3 parts (one of them may include binary message body), as defined by `decryptedSmpMessageBody` syntax:
+
+- `agentMsgHeader` - agent message header that contains sequential agent message ID for a particular SMP queue, agent timestamp (ISO8601) and the hash of the previous message.
+- `agentMessage` - a command/message to the other SMP agent:
+  - to establish the connection with two SMP queues (`helloMsg`, `replyQueueMsg`)
+  - to send and to acknowledge user messages (`clientMsg`, `acknowledgeMsg`)
+  - to notify another agent about queue deletion (`deleteQueueMsg`)
+- `msgPadding` - an optional message padding to make all SMP messages having consistent size as additional privacy protection.
+
+### Messages between SMP agents
+
+Message syntax below uses [ABNF][1].
+
+```abnf
+decryptedSmpMessageBody = agentMsgHeader CRLF agentMessage CRLF msgPadding
+agentMsgHeader = agentMsgId SP agentTimestamp SP previousMsgHash
+agentMsgId = 1*DIGIT ; sequential agent message ID set by the sending agent
+
+agentMessage = helloMsg / replyQueueMsg / deleteQueueMsg
+               / clientMsg / acknowledgeMsg
+
+msgPadding = *OCTET ; optional random bytes to get messages to the same size (as defined in SMP message size)
+
+helloMsg = %s"HELLO" SP signatureVerificationKey [SP %s"NO_ACK"]
+; NO_ACK means that acknowledgements to client messages will NOT be sent in this connection by the agent that sent `HELLO` message.
+signatureVerificationKey = encoded
```

Comment:
`encoded` - I still vote for abnf graph to be exhaustive in scope of a single syntax block

Reply:
at least `qInfo` explicitly refers to an outer syntax

Reply:
ok - added encoded - we can review it as a whole. It is actually quite common in RFCs to not have all syntax together. Maybe we could have a block with shared definitions and refer to it.